### ISSUE TITLE
Example "Select2 Drag and Drop Sorting" in Select2 4.0

### ISSUE THREAD

**Issue by Hemmils** [select2/select2#3004]
Hello Select2 gods,

I'm looking for the "Select2 Drag and Drop Sorting" example in Select2 4.0.
It's not on the example page. 

Is it still possible in 4.0?

**Comment by kevin-brown** [comment]
We have brought the old `onSortStart` and `onSortEnd` methods over to Select2 4.0. It wasn't used as often as most of the things we brought over, and I had no clue how it needed to be implemented.

If you happen to know what elements need to exist and how the sorting should happen, I'd be interested in making a decorator for the `selectionAdapter` that added support.

**Comment by Hemmils** [comment]
Thanks for the quick reply.
Sadly I got no clue. 
Maybe this is the problem: *The sorting is only available when Select2 is attached to a hidden input field.*
And there is no hidden input field in 4.0, right.

**Comment by kevin-brown** [comment]
> The sorting is only available when Select2 is attached to a hidden input field.

This mostly had to do with the fact that sorting a `<select>` requires moving the `<option>` elements around, which isn't always favorable. I still have no idea how I would want to handle this, though if you know of any other sortable dropdown plugins that we can look at, that would be useful.

**Comment by Hemmils** [comment]
Selectize.js has a drag_drop plugin, but it requires JQuery UI.
http://brianreavis.github.io/selectize.js/

**Comment by Dahkon** [comment]
What is the correct way to add sortable to select2 4.0 ?
I can find an exemple to make it work.

**Comment by andrewbaldock** [comment]
If you're trying to use jQueryUI Sortable per the [3.5.3 instructions](http://select2.github.io/select2/), this will make dragging work for Select2 v4:
```javascript
$("ul.select2-selection__rendered").sortable({
  containment: 'parent'
});
```
As mentioned above by @kevin-brown the select2 'value' accessed via select2('val') is not yet wired to be aware of the order change.  So until that is addressed you will need to roll your own 'sorted value getter', but it's doable via the dom.

[jsFiddle with examples](https://jsfiddle.net/andrewbaldock/ccj4twco/)

**Comment by WRidder** [comment]
I was wondering whether there are plans to natively support the custom ordering of a multiple select component?

**Comment by stevdom** [comment]
Hi,

I recently [a plugin](https://goodies.pixabay.com/jquery/tag-editor/demo.html) with the expected behavior, still I'd like to stick up with _select2_ rather than requiring X additional plugins for each use-case.

Has anybody come up with a solution based on _select2_ (with or without changes within base source)?

**Comment by lucasvg** [comment]
Me too. Is there any example of this done? I'm having such a hard time trying to accomplish this...

**Comment by joker-777** [comment]
This comment https://github.com/select2/select2/issues/1190#issuecomment-151832727 offers a possible solution. You only have to make sure that the options to choose from are ordered in the correct way initially.

**Comment by ambroiseRabier** [comment]
okay this work, i used it like this (whit jqueryUI sortable installed and require.js): 
```
    function select2_sortable($select2, pCallback){
        var select = $($select2);
        $(select).select2({
            width: '100%',
            createTag: function(params) {
                return undefined;
            }
        });
        var ul = $(select).next('.select2-container').first('ul.select2-selection__rendered');
        ul.sortable({
            placeholder : 'ui-state-highlight',
            forcePlaceholderSize: true,
            items       : 'li:not(.select2-search__field)',
            tolerance   : 'pointer',
            stop: function() {
                $($(ul).find('.select2-selection__choice').get().reverse()).each(function() {
                    var id = $(this).data('data').id;
                    var option = select.find('option[value="' + id + '"]')[0];
                    $(select).prepend(option);
                });
                pCallback();
            }
        });
    }
```

and then 
```
select2_sortable($("#YourSelectElement"), onOptionsChange);
```

**Comment by cfxd** [comment]
@ambroiseRabier you're a lifesaver, thank you!

I just had to remove the callback/param and this worked like a charm.

**Comment by litlife** [comment]
Method .first() does not accept any arguments.
I change code to:
`    function select2_sortable($select2, pCallback){
        var ul = $select2.next('.select2-container').find('ul.select2-selection__rendered');
        ul.sortable({
            placeholder : 'ui-state-highlight',
            forcePlaceholderSize: true,
            items       : 'li:not(.select2-search__field)',
            tolerance   : 'pointer',
            stop: function() {
                $($(ul).find('.select2-selection__choice').get().reverse()).each(function() {
                    var id = $(this).data('data').id;
                    var option = $select2.find('option[value="' + id + '"]')[0];
                    $select2.prepend(option);
                });
                pCallback();
            }
        });
    }`

**Comment by seaders** [comment]
`select2.find('option...')` doesn't seem to work with jQuery any more, either FYI.  You need to do *something like*,

`$(x_path + ' option').filter(function () { return $(this).html() === text; })`

I'm not using JQuery UI, anyway, I'm using this with JQuery EasyUI (a totally different beast!), so below is the drag'n'drop working with that.  It's not the most elegant, and I had to do a few things after delays because easyui seems to throw in arbitrary animations every now and again that I needed to wait for them to end, but it certainly works well enough.

```
function select2_sortable(select_id){
    var x_path = '#' + select_id;
    var $select = $(x_path);

    var ul_xpath = '#s2id_' + select_id + ' ul.select2-choices';
    $(ul_xpath).css({'min-height': '200px'});

    var li_xpath = ul_xpath + ' li';
    var $li = $(li_xpath);
    var $indicator = $('<li class="select2-search-choice indicator"> &nbsp; </li>').appendTo('body');

    $li.draggable({
        proxy:'clone',
        deltaX: 0,
        deltaY: 20,
        onStartDrag: function() {
            var proxy = $(this).draggable('proxy');
            proxy.css('z-index', 9999);
            $(this).hide();
        },
        onEndDrag: function() {
            endDrag($indicator, $(this));
        },
        onStopDrag: function() {
            endDrag($indicator, $(this));
        }
    });

    $li.droppable({
        accept: li_xpath,
        onDragOver:function(){
            $indicator.css({ display:'block' }).insertAfter(this);
        },
        onDragLeave:function(e, source){
            endDrag($indicator);
        },
        onDrop:function(e, source){
            $(source).insertAfter(this);
            endDrag($indicator, $(source));

            after_delay(function() {
                $.each($(li_xpath + ' div').get().reverse(), function () {
                    var text = $(this).text();
                    var $option = $(x_path + ' option').filter(function () {
                        return $(this).html() === text; });
                    $select.prepend($option);
                });
            });
        }
    });
}

function endDrag($indicator, $dragger) {
    $indicator.hide().appendTo('body');
    if($dragger) {
        $dragger.removeAttr('style');
        $dragger.show();
    }
}

function after_delay(fn) {
    setTimeout(fn, 10);
}
```

**Comment by pedrofurtado** [comment]
We don't have immediate plans to provide this. We are focused to fix some major UI bugs (that are majority of issues and PR's). But if you open a PR with unit tests, I will be glad to review and approve if everything is ok :+1:

**Comment by petr-panek** [comment]
I wrote a small extension for the Select2 Sortable.
This extension will change the order of the options in the "original" hidden select element after the sorting. It will also fire change event on the "original" hidden select element.

To use it, please initiate the select element this way:

     $('.rulePriceFromCompetitionShopNames').select2().setSelect2Sortable();

Select2 Sortable + Extension code:

    //jQuery Select2 Sortable
    //- enable select2 to be sortable via normal select element
    //- code updated for Select2 v4 (source: https://github.com/select2/select2/issues/1190#issuecomment-151832727)
    //author      : Vafour
    //inspired by : jQuery Chosen Sortable (https://github.com/mrhenry/jquery-chosen-sortable)
    //License     : GPL
    (function($){
    $.fn.extend({
        select2_sortable: function(){
            var select = $(this);
            $(select).select2({
                width: '100%',
                createTag: function(params) {
                    return undefined;
                }
            });
            var ul = $(select).next('.select2-container').first('ul.select2-selection__rendered');
            ul.sortable({
                placeholder : 'ui-state-highlight',
                forcePlaceholderSize: true,
                items       : 'li:not(.select2-search__field)',
                tolerance   : 'pointer',
                stop: function() {
                    $($(ul).find('.select2-selection__choice').get().reverse()).each(function() {
                        var id = $(this).data('data').id;
                        var option = select.find('option[value="' + id + '"]')[0];
                        $(select).prepend(option);
                    });
                }
            });
        }
    });
    }(jQuery));

    //jQuery Set Select2 Sortable (Extension to Select2 Sortable)
    //- This is extension will change the order of the options in the "original" hidden select element after the sorting.
    //- Will also fire change event on the "original" hidden select element.
    //author      : cenovyautomat.cz
    //License     : GPL
    (function( $ ){
    $.fn.setSelect2Sortable = function() {
        var originalSelectElement = this;   //ulož odkaz na originální select element
        $(originalSelectElement).next().children().children().children().sortable({containment: 'parent', stop: function( event, ui ) {  //ul element vytvořený Select2, děláme na něj sortable a definujeme funkci pro stop event
                $(originalSelectElement).next().children().children().children().children('[title]').each(function(){   //Select2 li elementy
                    var title = $(this).attr('title');
                    console.log(title);
                    var originalOptionElem =  $(originalSelectElement).children('[value="'+title+'"]');
                    originalOptionElem.detach();
                    $(originalSelectElement).append(originalOptionElem)
                });
                $(originalSelectElement).change();

            }
        });
    };
    })( jQuery );

(I'm not skilful JS developer, if someone'd like to rewrite the code better way it will be nice)

**Comment by danielbachhuber** [comment]
If your option label isn't the same as the ID, https://github.com/select2/select2/issues/3004#issuecomment-427919027 doesn't work.

Here's what I used, for an already instantiated `selectEl` element:

```js
var selectEl = $('select').select2();
selectEl.next().children().children().children().sortable({
	containment: 'parent', stop: function (event, ui) {
		ui.item.parent().children('[title]').each(function () {
			var title = $(this).attr('title');
			var original = $( 'option:contains(' + title + ')', selectEl ).first();
			original.detach();
			selectEl.append(original)
		});
		selectEl.change();
	}
});
```

**Comment by elshobokshy** [comment]
Thank you @danielbachhuber with some tweaking this works!

**Comment by JoomBall** [comment]
Small improvement to find the exact text


```
<option value="1">Hello Word</option>
<option value="2">Hello</option>
```

<pre>
var selectEl = $('select').select2({width: '100%'});

selectEl.next().find('ul').sortable({
	containment: 'parent', stop: function (event, ui) {
		ui.item.parent().children('[title]').each(function () {
			var title = $(this).attr('title');
			$('option:contains(\'' + title + '\')', selectEl).filter(function(){
				if ($(this).text() === title) {
					var original = $(this);
					original.detach();
					selectEl.append(original);
				}
			});
		});
		selectEl.change();
	}
});</pre>

**Comment by kamyargerami** [comment]
> Small improvement to find the exact text
> 
> ```
> <option value="1">Hello Word</option>
> <option value="2">Hello</option>
> ```
> 
> var selectEl = $('select').select2({width: '100%'});
> 
> selectEl.next().find('ul').sortable({
> 	containment: 'parent', stop: function (event, ui) {
> 		ui.item.parent().children('[title]').each(function () {
> 			var title = $(this).attr('title');
> 			$('option:contains(\'' + title + '\')', selectEl).filter(function(){
> 				if ($(this).text() === title) {
> 					var original = $(this);
> 					original.detach();
> 					selectEl.append(original);
> 				}
> 			});
> 		});
> 		selectEl.change();
> 	}
> });

Thank you. it works for my case perfectly :)))
